cpp/include/raft/matrix/detail/linewise_op.cuh

```diff
@@ -0,0 +1,364 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/cuda_utils.cuh>
+#include <raft/pow2_utils.cuh>
+#include <raft/vectorized.cuh>
+
+namespace raft {
+namespace matrix {
+namespace detail {
+
+template <typename Type, typename IdxType, std::size_t VecBytes, int BlockSize>
+struct Linewise {
+  static constexpr IdxType VecElems = VecBytes / sizeof(Type);
+
+  typedef raft::TxN_t<Type, VecElems> Vec;
+  typedef raft::Pow2<VecBytes> AlignBytes;
+  typedef raft::Pow2<VecElems> AlignElems;
+  typedef raft::Pow2<raft::WarpSize> AlignWarp;
+
+  template <typename Lambda, typename... Vecs>
+  static __device__ __forceinline__ void vectorCols(typename Vec::io_t* out,
+                                                    const typename Vec::io_t* in,
+                                                    const typename Vec::io_t* in_end,
+                                                    const IdxType rowLen,
+                                                    IdxType rowDiv,
+                                                    IdxType rowMod,
+                                                    Lambda op,
+                                                    Vecs... vecs) noexcept
+  {
+    constexpr IdxType warpPad = (AlignWarp::Value - 1) * VecElems;
+    Type args[sizeof...(Vecs)];
+    Vec v, w;
+    bool update = true;
+    for (; in < in_end; in += AlignWarp::Value, out += AlignWarp::Value, rowMod += warpPad) {
+      v.val.internal = __ldcv(in);
+      while (rowMod >= rowLen) {
+        rowMod -= rowLen;
+        rowDiv++;
+        update = true;
+      }
+      if (update) {
+        int l = 0;
+        ((args[l] = vecs[rowDiv], l++), ...);
+        update = false;
+      }
+#pragma unroll VecElems
+      for (int k = 0; k < VecElems; k++, rowMod++) {
+        if (rowMod == rowLen) {
+          rowMod = 0;
+          rowDiv++;
+          int l = 0;
+          ((args[l] = vecs[rowDiv], l++), ...);
+        }
```

Comment:
Hmm, not really. I tried all sorts of combinations of these. It would save at most two lines in total, at the danger of increasing the number of live registers...